presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateLongPercentileAggregations.java

```
@@ -35,49 +38,52 @@
+    public static void weightedInput(@AggregationState TDigestAndPercentileState state, @SqlType(StandardTypes.BIGINT) long value, @SqlType(StandardTypes.DOUBLE) double weight, @SqlType(StandardTypes.DOUBLE) double percentile)
     {
-        checkWeight(weight);
+        ApproximateDoublePercentileAggregations.weightedInput(state, toDoubleExact(value), weight, percentile);
+    }
 
-        QuantileDigest digest = state.getDigest();
+    @CombineFunction
+    public static void combine(@AggregationState TDigestAndPercentileState state, TDigestAndPercentileState otherState)
+    {
+        ApproximateDoublePercentileAggregations.combine(state, otherState);
+    }
 
-        if (digest == null) {
-            digest = new QuantileDigest(0.01);
-            state.setDigest(digest);
-            state.addMemoryUsage(digest.estimatedInMemorySizeInBytes());
+    @OutputFunction(StandardTypes.BIGINT)
+    public static void output(@AggregationState TDigestAndPercentileState state, BlockBuilder out)
+    {
+        TDigest digest = state.getDigest();
+        double percentile = state.getPercentile();
+        if (digest == null || digest.getCount() == 0.0) {
+            out.appendNull();
         }
+        else {
+            checkState(percentile != -1.0, "Percentile is missing");
+            checkCondition(0 <= percentile && percentile <= 1, INVALID_FUNCTION_ARGUMENT, "Percentile must be between 0 and 1");
+            BIGINT.writeLong(out, Math.round(digest.valueAt(percentile)));
+        }
+    }
 
-        state.addMemoryUsage(-digest.estimatedInMemorySizeInBytes());
-        digest.add(value, weight);
-        state.addMemoryUsage(digest.estimatedInMemorySizeInBytes());
-
-        // use last percentile
-        state.setPercentile(percentile);
+    private static double toDoubleExact(long value)
+    {
+        double doubleValue = (double) value;
+        checkState((long) doubleValue == value, "no exact double representation for long: %s", value);
```

This should throw a proper Presto exception with INVALID_FUNCTION_ARGUMENT